Temp save

henpe071 opened the pull request:
When creating a new project, everything is created in a temp folder. Clicking save will move the project to the assign folder.  Open project will also "open" the project in a temp folder. all changes made to the project are only saved in this temp folder and only recorded in the correct folder upon clicking save. 

When clicking save, a pop-up will appear if a project already exist in the assigned folder asking if you're sure you wanna overwrite the existing project. 

I might also have done some small cleanups in the code concerning some brackets and variable names.

henpe071 committed:
Vian now saves the project temporary before save is clicked

The project is temporarily saved in appdata-local-temp. When save project is clicked it is saved to the correct path.

henpe071 committed:
when clicking save the project in the temp folder gets copied to the correct path and saved there.

Also added some destructors to make the temporary folder go away ViAn is closed.

And i cleaned some things because it kinda bothered me.

henpe071 committed:
Now saves and opens correctly with thw temp map

Tho there's a seg fault when closing the program.

henpe071 committed:
Cleaned up the code and fixed a seg fault

Memory leaks everywhere!

fredrikiselius commented on ViAn/GUI/projectwidget.cpp:
```diff
@@ -582,14 +583,26 @@ void ProjectWidget::create_folder_item() {
     if (m_proj == nullptr ) return;
+    QDir dir(QString::fromStdString(m_proj->get_dir()));
+    if (dir.exists()) {
+        QMessageBox msg_box;
```
We've added quite alot of message boxes. We should refactor this type of code snippet into a function,

fredrikiselius commented on ViAn/Project/project.cpp:
```diff
@@ -4,17 +4,20 @@
     proj->load_saveable(full_path);
     // ensure changes to paths are saved
     proj->save_saveable(full_path);
+    if(proj->tmp_dir.isValid()){
```
And if its not? isValid() returns false if the temporary dir could not be created. This needs to be handled somehow. Wont it try to save to a nonexistent directory if it fails?

fredrikiselius commented on ViAn/Project/project.cpp:
```diff
@@ -25,10 +28,15 @@ Project* Project::fromFile(const std::string &full_path)
-    m_dir = dir_path + "/" + name + "/";
-    m_dir_bookmarks = m_dir + "Bookmarks/";
-    m_file = m_dir + name + ".vian";
+    if(tmp_dir.isValid()){
```
Same with this isValid()

fredrikiselius commented on ViAn/GUI/projectwidget.cpp:
```diff
@@ -45,6 +49,7 @@ ProjectWidget::ProjectWidget(QWidget *parent) : QTreeWidget(parent) {
 void ProjectWidget::new_project() {
     ProjectDialog* proj_dialog = new ProjectDialog();
     QObject::connect(proj_dialog, SIGNAL(project_path(QString, QString)), this, SLOT(add_project(QString, QString)));
+    delete proj_dialog;
```
The new project dialog is closed directly after it opens due to the delete here.

fredrikiselius commented on ViAn/GUI/projectwidget.cpp:
```diff
@@ -582,14 +583,26 @@ void ProjectWidget::create_folder_item() {
  */
 void ProjectWidget::save_project() {
     if (m_proj == nullptr ) return;
+    QDir dir(QString::fromStdString(m_proj->get_dir()));
```
This popup is triggered on each save. It should only happen when the project has never been saved to that location before.

fredrikiselius commented on ViAn/GUI/projectwidget.cpp:
```diff
@@ -582,14 +583,26 @@ void ProjectWidget::create_folder_item() {
     RecentProject rp;
     rp.load_recent();
-    rp.update_recent(m_proj->getName(), m_proj->full_path());
+    rp.update_recent(m_proj->get_name(), m_proj->get_file());
```
This no longer updates the recent projects list

fredrikiselius reviewed (changes requested):
The new additions broke both the new project window and recent project window for me. See code specific comments

henpe071 committed:
Remove a delete

henpe071 committed:
Fixed comments on github

Recent proj now creates its folder and have a max size of 10.
fixed a bug where remove not removing projects.
Added a seperate messagebox funktion in project widget.
Added more code in case temp_dir would be invalid.
Moved the "are you sure" save popup window to when creating a new project intead of asking on every save.

henpe071 committed:
Merge branch 'development' into Temp-save

henpe071 committed:
Fixed a problem with opening existing projects

henpe071 commented:
Now the things should be fixed and I also fixed a small thing I missed before.

fredrikiselius commented on ViAn/GUI/projectwidget.cpp:
```diff
@@ -60,11 +60,17 @@ void ProjectWidget::new_project() {
-    m_proj = new Project(name, path);
+    Project* new_proj = new Project(name, path);
+    if (!new_proj->tmp_dir_valid) {
+        delete m_proj;
```
Won't this put the program in a state without a project?

fredrikiselius commented on ViAn/Project/recentproject.cpp:
```diff
@@ -15,9 +15,10 @@ void RecentProject::update_recent(const std::string& name, const std::string &pr
     std::pair<std::string, std::string> new_proj = std::make_pair(name, project_path);
     auto item = std::find(recent_items.begin(), recent_items.end(), new_proj);
     if (item != recent_items.end()) recent_items.erase(item);
+    if (recent_items.size() >= 10) recent_items.resize(10);
```
10 should be made a constant

henpe071 committed:
Added a constant
